components/script/dom/blob.rs

```diff
@@ -173,7 +165,31 @@ impl Blob {
     /// Get a FileID representing the Blob content,
     /// used by URL.createObjectURL
     pub fn get_blob_url_id(&self) -> SelectedFileId {
-        match *self.blob_impl.borrow() {
+        let opt_sliced_parent = match *self.blob_impl.borrow() {
+            BlobImpl::Sliced(ref parent, ref rel_pos) => {
+                Some((parent.promote(/* set_valid is */ false), rel_pos.clone(), parent.Size()))
+            }
+            _ => None
+        };
+
+        match opt_sliced_parent {
+            Some((parent_id, rel_pos, size)) => self.create_sliced_url_id(&parent_id, &rel_pos, size),
+            None => self.promote(/* set_valid is */ true),
+        }
+    }
+
+    /// Promote non-Slice blob:
+    /// 1. Memory-based: The bytes in data slice will be transferred to file manager thread.
+    /// 2. File-based: Activation
+    /// Depending on set_valid, the returned FileID can be part of
+    /// valid or invalid Blob URL.
+    fn promote(&self, set_valid: bool) -> SelectedFileId {
```

Comment:
I moved part of logic in `get_blob_url_id` to `promote` due to the `DOMRefCell` borrow intricacies.